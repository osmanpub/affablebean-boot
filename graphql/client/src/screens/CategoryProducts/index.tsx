import React, { useEffect } from "react";
import { connect, useDispatch } from "react-redux";
import Footer from "../../components/Footer";
import Header from "../../components/Header";
import Products from "../../components/Products";
import { CategoryProducts as CategoryProductsState } from "../../interfaces/categories";
import { Match } from "../../interfaces/router";
import { fetchCategoryIfNeeded } from "../../net/category";
import { RootState } from "../../redux";

type Props = {
  category: CategoryProductsState;
  clearCart: Function;
  match: Match;
};

function CategoryProducts(props: Props) {
  const { category, match } = props;
  const { params } = match;
  const id = params.id;
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCategoryIfNeeded(id));
  }, [dispatch, id]);

  if (
    !category.category.hasOwnProperty("id") &&
    !category.category.hasOwnProperty("_id")
  ) {
    return null;
  }

  return (
    <div>
      <Header url={match.url} />
      <Products
        categories={category.categories}
        category={category.category}
        products={category.products}
      />
      <Footer />O
    </div>
  );
}

const mapStateToProps = (state: RootState) => {
  const { category } = state;

  return {
    category,
  };
};

export default connect(mapStateToProps)(CategoryProducts);
