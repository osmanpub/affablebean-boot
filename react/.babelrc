 
{
  "presets": ["@babel/preset-react"],
  "env": {
    "test": {
      "plugins": [
        ["istanbul", {
          "exclude": [
            "**/*.spec.js"
          ]
        }]
      ]
    }
  },
  "plugins": ["transform-class-properties", "istanbul"]
}
