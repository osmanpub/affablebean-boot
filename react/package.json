{
  "name": "affablebean",
  "version": "0.1.0",
  "private": true,
  "eslintConfig": {
    "extends": "react-app",
    "plugins": [
      "react-hooks"
    ],
    "rules": {
      "react-hooks/rules-of-hooks": "error",
      "react-hooks/exhaustive-deps": "warn"
    }
  },
  "scripts": {
    "build": "react-scripts build",
    "coveralls": "cat coverage/lcov.info | coveralls",
    "cypress:open": "cypress open",
    "cypress:run": "cypress run",
    "cypress:verify": "cypress verify",
    "eject": "react-scripts eject",
    "report:coverage": "nyc report --reporter=html",
    "report:coverage:text": "nyc report --reporter=text",
    "start": "node_modules/react-scripts/bin/react-scripts.js start &",
    "start:java": "cd .. && sh server && cd react && node_modules/react-scripts/bin/react-scripts.js start &",
    "test": "react-scripts test"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "dependencies": {
    "@reduxjs/toolkit": "^1.2.5",
    "axios": "^0.19.2",
    "node-rest-client": "^3.1.0",
    "react": "^16.12.0",
    "react-alert": "^6.0.0",
    "react-alert-template-basic": "^1.0.0",
    "react-dom": "^16.12.0",
    "react-hook-form": "^4.9.7",
    "react-redux": "^7.2.0",
    "react-router-dom": "^5.1.2",
    "react-scripts": "3.4.0",
    "redux": "^4.0.5",
    "styled-components": "^5.0.1"
  },
  "devDependencies": {
    "@babel/core": "^7.8.4",
    "@babel/preset-react": "^7.8.3",
    "@cypress/code-coverage": "^1.12.2",
    "@testing-library/jest-dom": "^4.2.4",
    "@testing-library/react": "^9.4.0",
    "@testing-library/user-event": "^7.2.1",
    "@types/axios": "^0.14.0",
    "@types/cypress": "^1.1.3",
    "@types/jest": "^25.1.3",
    "@types/react": "^16.9.21",
    "@types/react-alert": "^5.2.0",
    "@types/react-bootstrap": "^0.32.20",
    "@types/react-dom": "^16.9.5",
    "@types/react-redux": "^7.1.7",
    "@types/react-router-dom": "^5.1.3",
    "@types/react-test-renderer": "^16.9.2",
    "@types/redux": "^3.6.0",
    "@types/redux-mock-store": "^1.0.2",
    "@types/styled-components": "^5.0.0",
    "babel-plugin-istanbul": "^6.0.0",
    "babel-plugin-transform-class-properties": "^6.24.1",
    "coveralls": "^3.0.9",
    "cross-env": "^7.0.0",
    "cypress": "^4.0.2",
    "istanbul-lib-coverage": "^3.0.0",
    "nyc": "^15.0.0",
    "react-test-renderer": "^16.12.0",
    "redux-mock-store": "^1.5.4",
    "typescript": "^3.8.2"
  },
  "nyc": {
    "sourceMap": false,
    "instrument": false
  }
}
