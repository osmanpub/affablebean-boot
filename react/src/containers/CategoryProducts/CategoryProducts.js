import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Footer from "../../components/Footer";
import Header from "../../components/Header";
import Products from "../../components/Products";
import { fetchCategoryIfNeeded } from "../../net/category";

export class CategoryProducts extends Component {
  componentDidMount() {
    const { dispatch, match } = this.props;
    const { params } = match;

    dispatch(fetchCategoryIfNeeded(params.id));
  }

  componentDidUpdate(prevProps) {
    const { category, dispatch, match } = this.props;
    const { params } = match;
    const id = params.id;

    if (category.categories.length > 0 && Number(id) !== category.category.id) {
      dispatch(fetchCategoryIfNeeded(id));
    }
  }

  render() {
    const { cart, category, dispatch, match } = this.props;

    if (!category.category.hasOwnProperty("id")) {
      return null;
    }

    return (
      <div>
        <Header cart={cart} url={match.url} />
        <Products
          categories={category.categories}
          category={category.category}
          dispatch={dispatch}
          products={category.products}
        />
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = state => {
  const { cart, category } = state;

  return {
    cart,
    category
  };
};

export default connect(mapStateToProps)(CategoryProducts);

CategoryProducts.propTypes = {
  cart: PropTypes.object,
  category: PropTypes.object.isRequired,
  dispatch: PropTypes.func.isRequired,
  match: PropTypes.object.isRequired,
  url: PropTypes.string
};
