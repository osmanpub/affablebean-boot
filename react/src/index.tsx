import { configureStore } from "@reduxjs/toolkit";
import React, { lazy, Suspense } from "react";
import { render } from "react-dom";
import { Provider } from "react-redux";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import ErrorBoundary from "./components/ErrorBoundary";
import rootReducer from "./redux";
import * as serviceWorker from "./serviceWorker";
const Cart = lazy(() => import("./containers/Cart"));
const CategoryProducts = lazy(() => import("./containers/CategoryProducts"));
const Checkout = lazy(() => import("./containers/Checkout"));
const Contact = lazy(() => import("./containers/Contact"));
const Home = lazy(() => import("./containers/Home"));
const Privacy = lazy(() => import("./containers/Privacy"));

const store = configureStore({
  reducer: rootReducer
});

render(
  <React.StrictMode>
    <ErrorBoundary>
      <Provider store={store}>
        <Router>
          <Suspense fallback={<div>Loading...</div>}>
            <Switch>
              <Route path="/" exact component={Home} />
              <Route path="/category/:id" component={CategoryProducts} />
              <Route path="/checkout" component={Checkout} />
              <Route path="/contact" component={Contact} />
              <Route path="/privacy" component={Privacy} />
              <Route path="/viewCart/:clear" component={Cart} />
            </Switch>
          </Suspense>
        </Router>
      </Provider>
    </ErrorBoundary>
  </React.StrictMode>,
  document.getElementById("root")
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: http://bit.ly/CRA-PWA
serviceWorker.unregister();
