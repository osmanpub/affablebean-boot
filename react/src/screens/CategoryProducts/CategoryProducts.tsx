import React, { useEffect } from "react";
import { connect } from "react-redux";
import Footer from "../../components/Footer";
import Header from "../../components/Header";
import Products from "../../components/Products";
import { Cart } from "../../interfaces/cart";
import { CategoryProducts as CategoryProductsState } from "../../interfaces/categories";
import { fetchCategoryIfNeeded } from "../../net/category";

type Props = {
  cart: Cart;
  category: CategoryProductsState;
  clearCart: Function;
  dispatch: Function;
  match: any;
};

export function CategoryProducts(props: Props) {
  const { cart, category, dispatch, match } = props;
  const { params } = match;
  const id = params.id;

  useEffect(() => {
    dispatch(fetchCategoryIfNeeded(id));
  }, [category, dispatch, id]);

  if (!category.category.hasOwnProperty("id")) {
    return null;
  }

  return (
    <div>
      <Header cart={cart} url={match.url} />
      <Products
        cart={cart}
        categories={category.categories}
        category={category.category}
        dispatch={dispatch}
        products={category.products}
      />
      <Footer />
    </div>
  );
}

const mapStateToProps = (state: Props) => {
  const { cart, category } = state;

  return {
    cart,
    category
  };
};

export default connect(mapStateToProps)(CategoryProducts);
