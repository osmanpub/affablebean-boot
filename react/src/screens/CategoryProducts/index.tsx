import React, { useEffect } from "react";
import { connect } from "react-redux";
import Footer from "../../components/Footer";
import Header from "../../components/Header";
import Products from "../../components/Products";
import { CategoryProducts as CategoryProductsState } from "../../interfaces/categories";
import { Match } from "../../interfaces/router";
import { fetchCategoryIfNeeded } from "../../net/category";

type Props = {
  category: CategoryProductsState;
  clearCart: Function;
  dispatch: Function;
  match: Match;
};

function CategoryProducts(props: Props) {
  const { category, dispatch, match } = props;
  const { params } = match;
  const id = params.id;

  useEffect(() => {
    dispatch(fetchCategoryIfNeeded(id));
  }, [category, dispatch, id]);

  if (!category.category.hasOwnProperty("id")) {
    return null;
  }

  return (
    <div>
      <Header url={match.url} />
      <Products
        categories={category.categories}
        category={category.category}
        dispatch={dispatch}
        products={category.products}
      />
      <Footer />
    </div>
  );
}

const mapStateToProps = (state: Props) => {
  const { category } = state;

  return {
    category
  };
};

export default connect(mapStateToProps)(CategoryProducts);
