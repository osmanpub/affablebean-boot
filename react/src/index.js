import React from "react";
import { render } from "react-dom";
import { configureStore } from "redux-starter-kit";
import { Provider } from "react-redux";
import { BrowserRouter as Router, Route } from "react-router-dom";
import reducer from "./reducers";
import App from "./containers/App";
import CategoryProducts from "./containers/CategoryProducts";
import * as serviceWorker from "./serviceWorker";

const store = configureStore({
  reducer
});

render(
  // <React.StrictMode>
  <Provider store={store}>
    <Router>
      <Route path="/" exact component={App} />
      <Route path="/category/:id" component={CategoryProducts} />
    </Router>
  </Provider>,
  // </React.StrictMode>,
  document.getElementById("root")
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: http://bit.ly/CRA-PWA
serviceWorker.unregister();
