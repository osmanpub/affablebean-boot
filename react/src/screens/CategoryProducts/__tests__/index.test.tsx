import { act, fireEvent, render, waitForElement } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import CategoryProducts from "..";

const mockStore = configureStore([thunk]);

const initialState = {
  cart: {
    didInvalidate: false,
    isFetching: false,
    items: [],
    numberOfItems: 0,
    subtotal: 0
  },
  category: {
    categories: [
      {
        id: 1,
        name: "dairy"
      },
      {
        id: 2,
        name: "meats"
      },
      {
        id: 3,
        name: "bakery"
      },
      {
        id: 4,
        name: "fruit & veg"
      },
      {
        id: 5,
        name: "cereals"
      },
      {
        id: 6,
        name: "drinks"
      }
    ],
    category: {
      id: 1,
      name: "dairy"
    },
    didInvalidate: false,
    isFetching: false,
    products: [
      {
        id: 1,
        description: "semi skimmed (1L)",
        name: "milk",
        price: 1.7
      },
      {
        id: 2,
        description: "mild cheddar (330g)",
        name: "cheese",
        price: 2.39
      },
      {
        id: 3,
        description: "unsalted (250g)",
        name: "butter",
        price: 1.09
      },
      {
        id: 4,
        description: "medium-sized (6 eggs)",
        name: "free range eggs",
        price: 1.76
      }
    ]
  }
};

const store = mockStore(initialState);

const categoryProducts = (
  <Router>
    <Provider store={store}>
      <CategoryProducts
        clearCart={jest.fn()}
        match={{
          path: "/category/:id",
          url: "/category/1",
          isExact: true,
          params: { id: 1 }
        }}
      />
    </Provider>
  </Router>
);

describe("<CategoryProducts />", () => {
  it("renders correctly", () => {
    const component = renderer.create(categoryProducts).toJSON();
    // @ts-ignore
    expect(component).toMatchSnapshot();
  });

  it("show smaller navigatioal logo, on left", () => {
    const { getByAltText } = render(categoryProducts);
    const logo = getByAltText(/Affable Bean logo/i);
    // @ts-ignore
    expect(logo).toBeInTheDocument();
  });

  it("show larger logo, on right", () => {
    const { getByAltText } = render(categoryProducts);
    const logo = getByAltText(/the affable bean/i);
    // @ts-ignore
    expect(logo).toBeInTheDocument();
  });

  it("don't show view cart link", () => {
    const { getByText } = render(categoryProducts);
    const link = getByText(/view cart/i);
    // @ts-ignore
    expect(link).not.toBeVisible();
  });

  it("don't show proceed to checkout link", () => {
    const { getByText } = render(categoryProducts);
    const link = getByText(/proceed to checkout/i);
    // @ts-ignore
    expect(link).not.toBeVisible();
  });

  it("don't show cart header total", () => {
    const { getByTestId } = render(categoryProducts);
    const cart = getByTestId(/cart-hdr-total/i);
    // @ts-ignore
    expect(cart).not.toBeVisible();
  });

  it("show privacy", () => {
    const { getByTestId } = render(categoryProducts);
    const privacy = getByTestId(/footer-privacy/i);
    // @ts-ignore
    expect(privacy).toBeInTheDocument();
  });

  it("show contact", () => {
    const { getByTestId } = render(categoryProducts);
    const contact = getByTestId(/footer-contact/i);
    // @ts-ignore
    expect(contact).toBeInTheDocument();
  });
});

describe("<Products />", () => {
  it("selected category is dairy", () => {
    const { getByTestId } = render(categoryProducts);
    const selected = getByTestId(/selected-dairy/i);
    // @ts-ignore
    expect(selected).toHaveTextContent("dairy");
  });

  it("four products are shown", () => {
    const { getAllByTestId } = render(categoryProducts);
    const items = getAllByTestId(/product/i);
    // @ts-ignore
    expect(items).toHaveLength(4);
  });

  it("add 'milk' to cart", async () => {
    const { getAllByRole, getByAltText, getByTestId, getByText } = render(
      categoryProducts
    );
    const img = getByAltText(/milk/i);
    const milkUpdate = getAllByRole(/button/i)[0];

    act(() => {
      fireEvent.click(milkUpdate);
    });

    // @ts-ignore
    expect(img).toBeInTheDocument();

    const price = getByTestId(/price-milk/i);
    // @ts-ignore
    expect(price).toHaveTextContent("1.7");

    const link = await waitForElement(() => getByText(/view cart/i));
    // @ts-ignore
    // expect(link).toBeVisible(); // not working!
  });
});
