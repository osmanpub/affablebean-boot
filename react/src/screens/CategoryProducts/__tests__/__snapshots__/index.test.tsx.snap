// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryProducts /> renders correctly 1`] = `
<div>
  <div
    className="sc-AxgMl kehYEp"
  >
    <div
      className="sc-AxheI jbKZFY"
    >
      <div
        className="sc-AxiKw byMOZe"
      >
        <div
          className="sc-AxhCb kneLnw headerWidget"
        >
          <a
            className="bubble"
            data-cy="proceed-checkout"
            href="/checkout"
            onClick={[Function]}
          >
            proceed to checkout ➟
          </a>
        </div>
        <div
          className="sc-AxhUy dXnRMF headerWidget"
        >
          <span
            className="horizontalMargin"
            data-cy="cart-hdr-total"
            data-testid="cart-hdr-total"
          >
            <img
              alt="shopping cart icon"
              src="/static/img/cart.gif"
            />
             
            0 items
          </span>
          <a
            className="bubble"
            data-cy="cart-hdr-view-cart"
            href="/viewCart/false"
            onClick={[Function]}
          >
            view cart
          </a>
        </div>
      </div>
      <div
        className="sc-Axmtr cYyhem"
      >
        <a
          data-cy="home-logo"
          href="/"
          onClick={[Function]}
        >
          <img
            alt="Affable Bean logo"
            src="/static/img/logo.png"
          />
        </a>
      </div>
      <img
        alt="the affable bean"
        id="logoText"
        src="/static/img/logoText.png"
      />
    </div>
  </div>
  <div>
    <div
      className="sc-fzozJi fFXNmT"
    >
      <span
        className="sc-fzplWN hydICK"
        data-testid="selected-dairy"
      >
        <span
          className="categoryText"
        >
          dairy
        </span>
      </span>
      <a
        href="/category/2"
        onClick={[Function]}
      >
        <span
          className="categoryButton"
        >
          <span
            className="categoryText"
            data-cy="category-meats"
            data-testid="category-meats"
          >
            meats
          </span>
        </span>
      </a>
      <a
        href="/category/3"
        onClick={[Function]}
      >
        <span
          className="categoryButton"
        >
          <span
            className="categoryText"
            data-cy="category-bakery"
            data-testid="category-bakery"
          >
            bakery
          </span>
        </span>
      </a>
      <a
        href="/category/4"
        onClick={[Function]}
      >
        <span
          className="categoryButton"
        >
          <span
            className="categoryText"
            data-cy="category-fruit & veg"
            data-testid="category-fruit & veg"
          >
            fruit & veg
          </span>
        </span>
      </a>
      <a
        href="/category/5"
        onClick={[Function]}
      >
        <span
          className="categoryButton"
        >
          <span
            className="categoryText"
            data-cy="category-cereals"
            data-testid="category-cereals"
          >
            cereals
          </span>
        </span>
      </a>
      <a
        href="/category/6"
        onClick={[Function]}
      >
        <span
          className="categoryButton"
        >
          <span
            className="categoryText"
            data-cy="category-drinks"
            data-testid="category-drinks"
          >
            drinks
          </span>
        </span>
      </a>
    </div>
    <div
      className="sc-fzoLsD fTHuRQ"
    >
      <p
        className="sc-AxmLO hwiCmk"
      >
        dairy
      </p>
      <table
        className="sc-fzpans iXbeQw"
      >
        <tbody>
          <tr
            className="white"
            data-cy="product-milk"
            data-testid="product"
          >
            <td>
              <img
                alt="milk"
                src="/static/img/products/milk.png"
              />
            </td>
            <td
              data-testid="milk"
            >
              milk
              <br />
              <span
                className="smallText"
              >
                semi skimmed (1L)
              </span>
            </td>
            <td
              data-testid="price-milk"
            >
              € 
              1.70
            </td>
            <td>
                
            </td>
            <td>
              <button
                className="btn btn-primary btn-sm"
                onClick={[Function]}
              >
                add
              </button>
            </td>
            <td>
                
            </td>
          </tr>
          <tr
            className="lightBlue"
            data-cy="product-cheese"
            data-testid="product"
          >
            <td>
              <img
                alt="cheese"
                src="/static/img/products/cheese.png"
              />
            </td>
            <td
              data-testid="cheese"
            >
              cheese
              <br />
              <span
                className="smallText"
              >
                mild cheddar (330g)
              </span>
            </td>
            <td
              data-testid="price-cheese"
            >
              € 
              2.39
            </td>
            <td>
                
            </td>
            <td>
              <button
                className="btn btn-primary btn-sm"
                onClick={[Function]}
              >
                add
              </button>
            </td>
            <td>
                
            </td>
          </tr>
          <tr
            className="white"
            data-cy="product-butter"
            data-testid="product"
          >
            <td>
              <img
                alt="butter"
                src="/static/img/products/butter.png"
              />
            </td>
            <td
              data-testid="butter"
            >
              butter
              <br />
              <span
                className="smallText"
              >
                unsalted (250g)
              </span>
            </td>
            <td
              data-testid="price-butter"
            >
              € 
              1.09
            </td>
            <td>
                
            </td>
            <td>
              <button
                className="btn btn-primary btn-sm"
                onClick={[Function]}
              >
                add
              </button>
            </td>
            <td>
                
            </td>
          </tr>
          <tr
            className="lightBlue"
            data-cy="product-free range eggs"
            data-testid="product"
          >
            <td>
              <img
                alt="free range eggs"
                src="/static/img/products/free range eggs.png"
              />
            </td>
            <td
              data-testid="free range eggs"
            >
              free range eggs
              <br />
              <span
                className="smallText"
              >
                medium-sized (6 eggs)
              </span>
            </td>
            <td
              data-testid="price-free range eggs"
            >
              € 
              1.76
            </td>
            <td>
                
            </td>
            <td>
              <button
                className="btn btn-primary btn-sm"
                onClick={[Function]}
              >
                add
              </button>
            </td>
            <td>
                
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    className="sc-AxjAm jZHPdI"
  >
    <br />
    <span
      className="sc-AxirZ bSLrDO"
    >
      <a
        data-cy="privacy"
        data-testid="footer-privacy"
        href="/privacy"
        onClick={[Function]}
      >
        <button
          className="btn btn-link"
        >
          Privacy
        </button>
      </a>
      <a
        data-cy="contact"
        data-testid="footer-contact"
        href="/contact"
        onClick={[Function]}
      >
        <button
          className="btn btn-link"
        >
          Contact
        </button>
      </a>
       © 2020 the affable bean company
    </span>
  </div>
  O
</div>
`;
